mon,osd,mds: allow fs authorize to update caps

rishabh-d-dave opened the pull request:
Related PRs:  #42335, #46991, #46993, #46994, #47013, #50312, #50497, #50664, #50665, #50874, #50876, #50882, #51004, #51127 and #52008.

Modifies the code for `fs authorize` to update auth caps instead of exiting with an error when entity already has caps (whether or not for the FS). Very related case where this modification has an impact is that, unlike before, auth caps are updated even when the entity has no caps

Fixes: https://tracker.ceph.com/issues/47264



<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug
- [x] Add pending release note?

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

batrick commented:
@rishabh-d-dave I'll hold off on reviewing this because I suspect you want to do more work after the email discussion. Unless there's a particular change you want me to look at?

rishabh-d-dave commented:
@batrick 

If I could get a confirmation form you that I am moving in the right direction, it would be save me the effort of rewriting code later.

> @rishabh-d-dave I'll hold off on reviewing this because I suspect you want to do more work after the email discussion. Unless there's a particular change you want me to look at?

I am thinking to have a different ticket and PR for work related to the email discussion to keep the current patch set simple.

batrick commented on src/mon/AuthMonitor.cc:
```diff
@@ -1791,6 +1734,172 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+    all_caps_same = false;
+
+    if (cap_entity == "mon") {
+      if (ecap.find("fsname=" + fs) == string::npos) {
```
You're reimplementing cap parsing here. I'm not very familiar with this part of the Monitor but shouldn't it be possible to create an in-memory representation of the caps (like `MDSAuthCaps`), write a merge method for it (taking another `MDSAuthCaps`), and finally write its string representation out?

tchaikov replied:
instead of using raw string manipulation, would be better to expand `MDSCapParser` or define yet another parse using `boost::spirit`? so it's more readable and maintainable?

batrick reviewed (changes requested):
Otherwise it looks like you're heading in the right direction.

rishabh-d-dave commented:
@vshankar PTAL.

rishabh-d-dave commented:
jenkins test make check arm64

kotreshhr reviewed (commented):
The commit 4096b2ec982f50940f7526b0f251a49ef8326695 needs to be removed ?

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
Rebased and resolved a minor conflict.

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
The tests  too are now ready for review.

rishabh-d-dave commented:
If everything is fine, replacing regex by parser is the only thing left to go. https://github.com/ceph/ceph/pull/41779#discussion_r655828080

rishabh-d-dave commented:
Locally test_fs_authorize.py and testing_multifs_auth.py ran fine-ish. Testing with teuthology next.

rishabh-d-dave commented:
Both tests ran fine -
test_fs_authorize.py: http://pulpito.front.sepia.ceph.com/rishabh-2022-05-06_03:33:39-fs-wip-rishabh-fs-auth-subcmd-distro-basic-smithi
test_multifs_auth.py: http://pulpito.front.sepia.ceph.com/rishabh-2022-05-06_03:32:43-fs-wip-rishabh-fs-auth-subcmd-distro-basic-smithi

rishabh-d-dave commented:
Rebased to include the fix at PR #46169 since it affects execution of tests on this PR.

rishabh-d-dave commented:
The PR branch is rebased to include bug fixes written by me from latest master that made testing this PR difficult locally.

vshankar commented:
jenkins retest this please

rishabh-d-dave commented:
jenkins test windows

rishabh-d-dave commented:
jenkins test api

vshankar commented on qa/tasks/cephfs/caps_helper.py:
```diff
@@ -43,6 +43,15 @@ def write_test_files(self, mounts, testpath=''):
             dirpath = os_path_join(mount_x.hostfs_mntpt, testpath, dirname)
             mount_x.run_shell(f'mkdir {dirpath}')
             filepath = os_path_join(dirpath, filename)
+            # XXX: the reason behind adding filepathm, cephfs_name and both
```
I'm not 100% clean on how this catches bugs -- is there a verification of any sort somewhere?

rishabh-d-dave replied:
> I'm not 100% clean on how this catches bugs -- is there a verification of any sort somewhere?

Yes. There is a verification.

The problem is that if file data is also identical for all files created for testing across all CephFSs (say `cephfs1` and `cephfs2` in this case) it's possible for a test case meant for `cephfs1` to pass even when it's looking at a file from `cephfs2` (instead of `cephfs1`).

The verification is that code meant for testing access levels (`r` and `rw`) look at the file data just obtained from read_file() and at the file data saved in `filedata` to make sure that it is looking at right file. Copying from `conduct_pos_test_for_read_caps()` and `conduct_pos_test_for_write_caps()` -

```
self.assertEqual(data, contents) # data is from filedata and contents is obtained from read_file()
```

vshankar replied:
OK - makes sense.

vshankar commented on qa/tasks/cephfs/caps_helper.py:
```diff
@@ -60,29 +60,49 @@ def run_cap_tests(self, perm, mntpt=None):
+        while moncap.find('fsname=') != -1:
+            fsname_first_char = moncap.index('fsname=') + len('fsname=')
+
+            if ',' in moncap:
```
perhaps use regex?

vshankar reviewed (changes requested):
@rishabh-d-dave Many commit, esp those in qa/cephfs, could possibly be standalone changes. E.g., commit d467151f37336a720841f4d3fafd823d5a693436 and df5b2d8b401294c1d4e0220a2332ff84d73fec08 (and others).

I suggest separate PRs for these commits to avoid keeping them blocked from merging since this PR would possibly go through a few more review iterations.

rishabh-d-dave commented on qa/tasks/cephfs/caps_helper.py:
```diff
@@ -95,10 +95,9 @@ def run_mon_cap_tests(self, def_fs, moncap, client_id, keyring):
-        # in MON/MDS caps
-        if 'fsname' not in moncap:
-            self.assertIn(self.fs.name, fsls)
+        if 'fsname=' not in moncap:
```
Yes. I expect this commit to disappear from this PR branch with a rebase when that PR merges.

rishabh-d-dave commented on qa/tasks/cephfs/caps_helper.py:
```diff
@@ -68,6 +68,7 @@ def conduct_neg_test_for_write_caps(self, filepaths, mounts):
                     cmdargs.append(path)
                     mount.negtestcmd(args=cmdargs, retval=1,
                                      errmsg='permission denied')
+                    cmdargs.pop(len(cmdargs)-1)
```
True, that'll be better.

rishabh-d-dave replied:
Fixed.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1801,6 +1764,29 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+}
+
+void AuthMonitor::_print_auth(EntityName& entity, EntityAuth& eauth,
+                              bufferlist& rdata, Formatter* fmtr,bool just_key)
```
> minor nit: space after ", bool..."

IIRC I did intentionally to save it from going over 80 chars.

vshankar replied:
that's fine - let's keep the param spacing as it should be.

rishabh-d-dave commented on qa/tasks/cephfs/caps_helper.py:
```diff
@@ -7,35 +7,56 @@
+        needs to be tested, call run_cap_tests(), run_mon_cap_tests() or
+        run_mds_cap_tests() as per the need.
+        """
+        self.mounts = mounts
```
Strange, I can't find this line in caps_helper.py. Search from the comment above, I see this -

```
        needs to be tested, call run_cap_tests(), run_mon_cap_tests() or
        run_mds_cap_tests() as per the need.
        """
        dirname, filename = 'testdir', 'testfile'
        self.test_set = []
```

rishabh-d-dave replied:
Okay, I see it in commit 3f4983a597787916bda717c1858bc994ea049c5b.

Not sure. Perhaps, yes. I'll keep this in mind and try this change.

vshankar commented on qa/tasks/cephfs/caps_helper.py:
```diff
@@ -1,12 +1,32 @@
+        filepaths, filedata = [], 'testdata'
+        dirname, filename = 'testdir', 'testfile'
+
+        for mount_x in mounts:
```
That brings out the best out of Python, doesn't it? :)

Anyway, let use `for mnt in mounts`?

vshankar commented on qa/tasks/ceph_manager.py:
```diff
@@ -1614,6 +1614,23 @@ def raw_cluster_cmd_result(self, *args, **kwargs):
+        if client_id.find('client.') != -1:
+            client_id = client_id.replace('client.', '')
+
+        keyring = self.run_cluster_cmd(args=f'auth get client.{client_id}',
```
I meant first converting client.foo -> foo and later doing "auth get client.foo".

rishabh-d-dave replied:
The idea is to make sure that this method works fine even when full entity name (client.foo) is received instead of just the ID (foo).

rishabh-d-dave commented:
I'm rerunning tests with teuth once more before I rebase.

rishabh-d-dave commented:
Tests ran fine - http://pulpito.front.sepia.ceph.com/rishabh-2022-06-08_05:47:38-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

vshankar commented:
@rishabh-d-dave is this ready for another round of review?

rishabh-d-dave commented:
Most tests pased. 

test_admin.py also exposes the bug I found out last week. I've ran test_admin.py twice -- first time as it is and second time after blocking the tests that exposed the bug (to make sure that rest of the tests also get executed).

test_admin:
passed, because failing test were blocked: http://pulpito.front.sepia.ceph.com/rishabh-2022-06-28_14:25:16-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/
failed: http://pulpito.front.sepia.ceph.com:80/rishabh-2022-06-28_12:39:40-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

test_fs_authorize: http://pulpito.front.sepia.ceph.com:80/rishabh-2022-06-28_16:08:10-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

test_multifs_auth: http://pulpito.front.sepia.ceph.com/rishabh-2022-06-28_10:34:33-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi

rishabh-d-dave commented:
http://pulpito.front.sepia.ceph.com/rishabh-2022-06-28_18:17:42-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2022-06-28_18:16:03-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2022-06-28_18:13:30-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

rishabh-d-dave commented on qa/tasks/vstart_runner.py:
```diff
@@ -761,6 +761,23 @@ def find_remote(self, daemon_type, daemon_id):
+        if client_id.find('client.') != -1:
+            client_id = client_id.replace('client.', '')
+
+        keyring = self.run_cluster_cmd(args=f'auth get client.{client_id}',
```
The idea is to make sure that this method works fine even when full entity name (client.foo) is received instead of just the ID (foo).

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
jenkins test windows

rishabh-d-dave commented:
Got expected results -

test_multifs_auth: http://pulpito.front.sepia.ceph.com/rishabh-2022-07-01_03:23:16-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

test_fs_authorize.py: http://pulpito.front.sepia.ceph.com/rishabh-2022-06-30_18:39:59-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

test_admin.py. success: http://pulpito.front.sepia.ceph.com/rishabh-2022-06-30_18:39:59-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

test_admin.py, failed: http://pulpito.front.sepia.ceph.com/rishabh-2022-06-30_18:43:07-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

vshankar commented:
@rishabh-d-dave Did you post and update for further reviews or otherwise? I see many of my comments are not yet addressed.

rishabh-d-dave commented:
> @rishabh-d-dave Did you post and update for further reviews or otherwise? I see many of my comments are not yet addressed.

Yes, for now, I've left out the C++  requested changes. I was figuring out if separating (most of) QA changes would be beneficial or complicating. I'll move a lot of commits from this PR to a separate PR ASAP.

rishabh-d-dave commented:
All test ran fine, including the ones that were blocked before (the bug has been fixed) -

http://pulpito.front.sepia.ceph.com/rishabh-2022-07-09_04:22:23-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi
http://pulpito.front.sepia.ceph.com/rishabh-2022-07-09_04:22:15-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi
http://pulpito.front.sepia.ceph.com/rishabh-2022-07-09_04:22:04-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi

rishabh-d-dave commented:
jenkins test windows

rishabh-d-dave commented:
jenkins test windows

rishabh-d-dave commented:
Retested to make sure rebase was fine -

test_fs_authorize.py: http://pulpito.front.sepia.ceph.com/rishabh-2022-07-20_14:57:17-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/
test_multifs_auth.py: http://pulpito.front.sepia.ceph.com/rishabh-2022-07-20_14:55:51-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/
test_admin.py: http://pulpito.front.sepia.ceph.com/rishabh-2022-07-20_15:04:38-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
Reran all test for this PR -

qa.tasks.cephfs.admin.test_fs_authorize: http://pulpito.front.sepia.ceph.com/rishabh-2022-11-20_11:20:51-fs-wip-rishabh-fs-auth-subcmd-20Nov2022-distro-default-smithi
All jobs passed

qa.tasks.cephfs.test_admin: http://pulpito.front.sepia.ceph.com/rishabh-2022-11-20_21:06:29-fs-wip-rishabh-fs-auth-subcmd-20Nov2022-distro-default-smithi
2/6 jobs failed, failures are unrelated.

qa.tasks.cephfs.test_multifs_auth: http://pulpito.front.sepia.ceph.com/rishabh-2022-11-20_11:18:58-fs-wip-rishabh-fs-auth-subcmd-20Nov2022-distro-default-smithi/
4/12 jobs failed, failure are unrelated

rzarzynski commented:
@vshankar: are you fine with the changes?

@rishabh-d-dave: should be it rested since the last last time? How severe are the last changes? It looks this PR should go also the `rados` suite.

vshankar commented:
> @vshankar: are you fine with the changes?

I looked at it a while back and there was some pending work too be finished before this can be tested. I haven't looked since then (there are other priority items for reef).

> 
> @rishabh-d-dave: should be it rested since the last last time? How severe are the last changes? It looks this PR should go also the `rados` suite.

vshankar commented:
Discussed with @rishabh-d-dave - pending item is to use boost:regex rather than writing a custom parser (and a round of review).

rishabh-d-dave commented:
> Discussed with @rishabh-d-dave - pending item is to use boost:regex rather than writing a custom parser (and a round of review).

The pending task is to use boost::spirit instead of std::regex.

rishabh-d-dave commented:
Just a rebase.

rishabh-d-dave commented:
Tests for current version PR branch -
PASSED
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-03_17:16:19-fs:multifs-wip-rishabh-fs-auth-subcmd-2023Mar01-1118-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-03_17:16:38-fs:functional-wip-rishabh-fs-auth-subcmd-2023Mar01-1118-distro-default-smithi/

FAILED
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-03_17:16:48-fs:functional-wip-rishabh-fs-auth-subcmd-2023Mar01-1118-distro-default-smithi/

Tests for previous version of PR branch. All tests passed -
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-03_17:19:21-fs:multifs-wip-rishabh-fs-auth-subcmd-2023Feb28-2023Mar02-1948-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-03_17:19:39-fs:functional-wip-rishabh-fs-auth-subcmd-2023Feb28-2023Mar02-1948-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-03_17:19:48-fs:functional-wip-rishabh-fs-auth-subcmd-2023Feb28-2023Mar02-1948-distro-default-smithi/

rishabh-d-dave commented:
Bug fix for the failing job reported [here](https://github.com/ceph/ceph/pull/41779#issuecomment-1456833881) and few minor adjustments.

rishabh-d-dave commented:
Bug's fixed, tests ran fine this time. The only failing job failed for unrelated reasons.

test_fs_authorize.py: http://pulpito.front.sepia.ceph.com/rishabh-2023-03-07_03:31:17-fs:functional-wip-rishabh-fs-auth-subcmd-2023Feb28-2023Mar02-1948-distro-default-smithi/
test_admin.py: http://pulpito.front.sepia.ceph.com/rishabh-2023-03-07_03:31:08-fs:functional-wip-rishabh-fs-auth-subcmd-2023Feb28-2023Mar02-1948-distro-default-smithi/
test_multifs_auth.py: http://pulpito.front.sepia.ceph.com/rishabh-2023-03-07_03:30:48-fs:multifs-wip-rishabh-fs-auth-subcmd-2023Feb28-2023Mar02-1948-distro-default-smithi/

rishabh-d-dave commented:
Test results for current version of PR branch - 
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-07_11:39:41-fs:functional-wip-rishabh-fs-auth-subcmd-3-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-07_11:39:32-fs:functional-wip-rishabh-fs-auth-subcmd-3-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-07_11:39:13-fs:multifs-wip-rishabh-fs-auth-subcmd-3-distro-default-smithi/

Lots of failure but all failures are unrelated.

rishabh-d-dave commented:
Re-ran tests for current version of PR branch. Tests passed this time, unrelated failure did not occur again -
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-07_20:32:02-fs:multifs-wip-rishabh-fs-auth-subcmd-3-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-07_20:32:19-fs:functional-wip-rishabh-fs-auth-subcmd-3-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-07_20:32:27-fs:functional-wip-rishabh-fs-auth-subcmd-3-distro-default-smithi/

rishabh-d-dave commented:
Merged the commit that replaced std::regex code with Boost.Spirit code.

rishabh-d-dave commented:
jenkins test make check arm64

rishabh-d-dave commented:
jenkins test make check arm64

rishabh-d-dave commented:
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-08_19:29:09-fs:multifs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-08_19:29:50-fs:functional-wip-rishabh-fs-auth-subcmd-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-08_19:29:59-fs:functional-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

rishabh-d-dave commented:
Failures looks unrelated - https://jenkins.ceph.com/job/ceph-pull-requests/111897/.

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
Tests passed with latest changes -
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-09_15:07:58-fs:multifs-wip-rishabh-fs-auth-subcmd-printauth-distro-default-smithi
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-09_15:08:18-fs:functional-wip-rishabh-fs-auth-subcmd-printauth-distro-default-smithi
http://pulpito.front.sepia.ceph.com/rishabh-2023-03-09_15:08:27-fs:functional-wip-rishabh-fs-auth-subcmd-printauth-distro-default-smithi

rishabh-d-dave commented:
@vshankar @batrick PTAL.

vshankar commented:
> @vshankar Since most of it looks okay, I'll go ahead for the refactoring to replace regex by Boost.Spirit.

@rishabh-d-dave was this implemented? Can you point me to the diff that uses this so that I do not have to go over large parts of the change again.

rishabh-d-dave commented:
@vshankar 
> > @vshankar Since most of it looks okay, I'll go ahead for the refactoring to replace regex by Boost.Spirit.
> 
> @rishabh-d-dave was this implemented? Can you point me to the diff that uses this so that I do not have to go over large parts of the change again.

Yes. This has been implemented. The diff was supposed to be somewhere after this comment - https://github.com/ceph/ceph/pull/41779#issuecomment-1449374185. But it's not present. Most of the `diff`s after this comment are changes I made to this PR _after_ replacing regex code by Boost.Spirit code. Strange. I'll point to  the changes myself -

vshankar commented on src/mon/AuthMonitor.cc:
```diff
@@ -1664,6 +1664,41 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
       + " tag " + pg_pool_t::APPLICATION_NAME_CEPHFS
       + " data=" + filesystem;
 
+    vector<string> newcaps = {"mon", mon_cap_string, "osd", osd_cap_string,
```
@rishabh-d-dave what about this comment? I don't a follow up discussion on this. Or am I looking at an older commit? This change was pointed from comment: https://github.com/ceph/ceph/pull/41779#issuecomment-1471543890

vshankar commented:
> @vshankar
> 
> > > @vshankar Since most of it looks okay, I'll go ahead for the refactoring to replace regex by Boost.Spirit.
> > 
> > 
> > @rishabh-d-dave was this implemented? Can you point me to the diff that uses this so that I do not have to go over large parts of the change again.
> 
> Yes. This has been implemented. The diff was supposed to be somewhere after this comment - [#41779 (comment)](https://github.com/ceph/ceph/pull/41779#issuecomment-1449374185). But it's not present. Most of the `diff`s after this comment are changes I made to this PR _after_ replacing regex code by Boost.Spirit code. Strange. I'll point to the changes myself -
> 
> Look at all the changes in MonCap.cc, MonCap.h, OSDCap.cc, OSDCap.h, MDSAuthCaps.cc and MDSAuthCap.h. In AuthMonitor.cc,
> 
> * review this whole method - https://github.com/ceph/ceph/pull/41779/files#diff-53eecdfcf56b08f0ec4b07dfb7eb394518f408e862c4a0173e7d40ebefb204e4R1884
> * and review this small block of code - https://github.com/ceph/ceph/pull/41779/files#diff-53eecdfcf56b08f0ec4b07dfb7eb394518f408e862c4a0173e7d40ebefb204e4R1803-R1807

The above links have _no_ use of boost spirit? Did you point to the incorrect commit(s)?

> 
> For changes caused in tests due to this change in C++ code, see changes in test_fs_authorize.py and caps_helper.py from this diff - https://github.com/ceph/ceph/compare/044709f352a6ad090e5e3f38970cb68b58d28a04..d50fb4ceed55c5048ca9ef31b677757ef901d906
> 
> Besides the change caused by replacing regex code by Boost.Spirit code

I'm seeing all changed file here: https://github.com/ceph/ceph/pull/41779/files

I cannot find the changes that use boost spirit.

> I caught some bugs while reading my own code.

Are you talking about the boost changes? Could you point me to those?

> Past review and tests didn't catch this bug.

Which bug? Bugs which were caused when switching to boost? I never reviewed those since I never found those changes.

> IMO, it would be easier to review this commit instead of reviewing the diff for these changes - [ad94515](https://github.com/ceph/ceph/commit/ad94515061e5dbf84db9b472560165836127affd)
> 
> Changes in AuthMonitor.h is tiny. So review it entirely.

Either I'm confused or github diff is acting crazy to _not_ show me the changes which involve using boost.

rishabh-d-dave commented:
> > @vshankar
> > > > @vshankar Since most of it looks okay, I'll go ahead for the refactoring to replace regex by Boost.Spirit.
> > > 
> > > 
> > > @rishabh-d-dave was this implemented? Can you point me to the diff that uses this so that I do not have to go over large parts of the change again.
> > 
> > 
> > Yes. This has been implemented. The diff was supposed to be somewhere after this comment - [#41779 (comment)](https://github.com/ceph/ceph/pull/41779#issuecomment-1449374185). But it's not present. Most of the `diff`s after this comment are changes I made to this PR _after_ replacing regex code by Boost.Spirit code. Strange. I'll point to the changes myself -
> > Look at all the changes in MonCap.cc, MonCap.h, OSDCap.cc, OSDCap.h, MDSAuthCaps.cc and MDSAuthCap.h. In AuthMonitor.cc,
> > 
> > * review this whole method - https://github.com/ceph/ceph/pull/41779/files#diff-53eecdfcf56b08f0ec4b07dfb7eb394518f408e862c4a0173e7d40ebefb204e4R1884
> > * and review this small block of code - https://github.com/ceph/ceph/pull/41779/files#diff-53eecdfcf56b08f0ec4b07dfb7eb394518f408e862c4a0173e7d40ebefb204e4R1803-R1807
> 
> The above links have _no_ use of boost spirit? Did you point to the incorrect commit(s)?
> 
> > For changes caused in tests due to this change in C++ code, see changes in test_fs_authorize.py and caps_helper.py from this diff - https://github.com/ceph/ceph/compare/044709f352a6ad090e5e3f38970cb68b58d28a04..d50fb4ceed55c5048ca9ef31b677757ef901d906
> > Besides the change caused by replacing regex code by Boost.Spirit code
> 
> I'm seeing all changed file here: https://github.com/ceph/ceph/pull/41779/files
> 
> I cannot find the changes that use boost spirit.
> 
> > I caught some bugs while reading my own code.
> 
> Are you talking about the boost changes? Could you point me to those?
> 
> > Past review and tests didn't catch this bug.
> 
> Which bug? Bugs which were caused when switching to boost? I never reviewed those since I never found those changes.
> 
> > IMO, it would be easier to review this commit instead of reviewing the diff for these changes - [ad94515](https://github.com/ceph/ceph/commit/ad94515061e5dbf84db9b472560165836127affd)
> > Changes in AuthMonitor.h is tiny. So review it entirely.
> 
> Either I'm confused or github diff is acting crazy to _not_ show me the changes which involve using boost.

No, The work that was done by regex code in last version of this PR is now done by Boost.Spirit code but I didn't have to add any Boost.Spirit code for parsing caps, existing parsers sufficed for this. To be precise, the work that was done by regex code is now done by merge() methods in MonCap, OSDCap and MDSAuthCaps. I've linked all these changes in my previous comment.

rishabh-d-dave commented:
@vshankar 
> > Past review and tests didn't catch this bug.
> 
> Which bug? Bugs which were caused when switching to boost? I never reviewed those since I never found those changes.

The bugs that originated from code besides that directly handled caps string code.
 
> > IMO, it would be easier to review this commit instead of reviewing the diff for these changes - [ad94515](https://github.com/ceph/ceph/commit/ad94515061e5dbf84db9b472560165836127affd)
> > Changes in AuthMonitor.h is tiny. So review it entirely.
> 
> Either I'm confused or github diff is acting crazy to _not_ show me the changes which involve using boost.

No, The work that was done by regex code in last version of this PR is now done by Boost.Spirit code but I didn't have to add any Boost.Spirit code for parsing caps, existing parsers sufficed for this. To be precise, the work that was done by regex code is now done by merge() methods in MonCap, OSDCap and MDSAuthCaps. I've linked all these changes in my previous comment.

vshankar commented on src/mon/AuthMonitor.cc:
```diff
@@ -1929,6 +1876,145 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+
+    if (cap_entity == "mon") {
+      MonCap c_cap, n_cap;
+      if (not c_cap.parse(c_cap_str, out))
```
nit: enclose within `{}`

rishabh-d-dave replied:
The code under `if` was a single line, so it wasn't needed.

In latest push, it has multiple line, so it's enclosed now.

vshankar replied:
> The code under `if` was a single line, so it wasn't needed.

Convention is to enclose even a single line within `{}`

vshankar commented on src/mon/AuthMonitor.cc:
```diff
@@ -1929,6 +1876,145 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+      if (not c_cap.parse(c_cap_str, out))
+	*out << "error parsing " << cap_entity << "caps client already holds";
+      if (not n_cap.parse(n_cap_str, out))
+	*out << "error parsing new " << cap_entity << "caps";
```
Shouldn't this error out rather than continuing and calling `merge`?

rishabh-d-dave replied:
True. Nice catch. Thanks! Fixed.

vshankar commented on src/mon/AuthMonitor.cc:
```diff
@@ -1929,6 +1876,145 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+	n_cap_str = c_cap_str;
+      }
+    } else if (cap_entity == "mds") {
+      MDSAuthCaps c_cap, n_cap;
+      if (not c_cap.parse(g_ceph_context, c_cap_str, out))
+	*out << "error parsing " << cap_entity << "caps client already holds";
+      if (not n_cap.parse(g_ceph_context, n_cap_str, out))
+	*out << "error parsing new " << cap_entity << "caps";
+
+      if (c_cap.merge(n_cap)) {
+	all_caps_same = false;
+	n_cap_str = c_cap.str();
+      } else {
+	n_cap_str = c_cap_str;
+      }
```
This code duplication can be done away with by having the respective subsystem auth caps derive from an abstract class and then calling the respective `parse` and `merge` virtual functions. WDYT?

rishabh-d-dave replied:
True. Easier thing to do is not pass g_ceph_context as it is globally available. Then the interface of all three cap classes would be the same and all these repetition will be not needed. 

Here's the PR for this - https://github.com/ceph/ceph/pull/50312. I talked about it in our last meeting and requested a review for it.

vshankar replied:
> True. Easier thing to do is not pass g_ceph_context as it is globally available. Then the interface of all three cap classes would be the same and all these repetition will be not needed.

Could you elaborate a bit more. How does not passing g_ceph_context help with parse/merge for auth caps?

> 
> Here's the PR for this - #50312. I talked about it in our last meeting and requested a review for it.

rishabh-d-dave replied:
The interfaces for all 3 would be the same, same call would work for all three.

vshankar replied:
In that case, we should take this approach. And, I think we could invoke the relevant routine using typecasting.

rishabh-d-dave replied:
Yes. Once that PR is merged, I'll change this code accordingly. I was expecting it too happen sooner but reviews as usual are taking time.

vshankar replied:
> Yes. Once that PR is merged, I'll change this code accordingly. I was expecting it too happen sooner but reviews as usual are taking time.

I added `needs-qa` to that PR.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1929,6 +1876,145 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+    }
+
+    string c_cap_str; // current cap held by entity's auth keyring
+    auto i = e_auth.caps[cap_entity].cbegin();
```
`it` is already used as the loop variable.

vshankar replied:
use iter.

rishabh-d-dave replied:
Changed.

rishabh-d-dave replied:
BTW, what's wrong with `i`?

vshankar replied:
readability - it and iter _sound_ like iterators.

rishabh-d-dave replied:
Okay.

rishabh-d-dave commented on src/mds/MDSAuthCaps.cc:
```diff
@@ -373,6 +373,84 @@ bool MDSAuthCaps::parse(CephContext *c, std::string_view str, ostream *err)
+    /* TODO: check case where cap is "allow rw *". */
+
+    if (g.match.fs_name == ng.match.fs_name) {
+      if (g.match.path == ng.match.path and
```
Okay, why do we prefer `&&` over `and`?

rishabh-d-dave replied:
Done.

vshankar replied:
> Okay, why do we prefer `&&` over `and`?

Personally, I'd always use `&&` - old habits. Moreover, we haven't been using `and` is ceph C++ sources.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1929,6 +1876,145 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+
+  for (auto it = newcaps.begin(); it != newcaps.end(); it += 2) {
+    string& cap_entity = *it;
+    string& n_cap_str = *(it + 1); // new cap to be added to the current cap
```
That's not possible because the vector is created in the same patch to have 6 elements.

rishabh-d-dave replied:
IMO adding assert is redundant because it has been hardcoded to have 6 members. Copying from rishabh-d-dave:ceph src/mon/AuthMonitor.cc
```
    vector<string> newcaps = {"mon", mon_cap_string, "osd", osd_cap_string,
			      "mds", mds_cap_string};
```

vshankar replied:
In case this is used in code elsewhere. Also, what about comment: https://github.com/ceph/ceph/pull/41779/commits/e2f4e203a0c9b7687bf7577f8ed7989b2ff7c1b9#r885461143

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1929,6 +1876,145 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+}
+
+void AuthMonitor::_print_auth(EntityName& entity, EntityAuth& eauth,
+  bufferlist& rdata, Formatter* fmtr, bool just_key, bool pending_key,
```
Done.

rishabh-d-dave replied:
Apply `const` is not possible because underlying methods do not do the same.

vshankar replied:
ah ok - items are encoded in bufferlist in this routine.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1929,6 +1876,145 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+}
+
+void AuthMonitor::_print_key(EntityName& entity, EntityAuth& eauth,
+  bufferlist& rdata, Formatter* fmtr, bool pending_key,
```
Done.

rishabh-d-dave replied:
Same as https://github.com/ceph/ceph/pull/41779#discussion_r1141995701.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1929,6 +1876,145 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+ * caps already present in the client's auth keyring with the new caps
+ * provided by the user at "fs authorize" command.
+ */
+bool AuthMonitor::_gen_wanted_caps(EntityAuth& e_auth,
```
I've made this change

rishabh-d-dave replied:
Looks like it not possible to use `operator []` when the variable has `const` qualifier applied to it.

vshankar replied:
use `.at()` then?

rishabh-d-dave commented on src/mds/MDSAuthCaps.cc:
```diff
@@ -373,6 +373,84 @@ bool MDSAuthCaps::parse(CephContext *c, std::string_view str, ostream *err)
+  }
+
+  str += " fsname=" + match.fs_name;
+  if (not match.path.empty())
```
Done.

vshankar commented on src/mds/MDSAuthCaps.cc:
```diff
@@ -373,6 +373,84 @@ bool MDSAuthCaps::parse(CephContext *c, std::string_view str, ostream *err)
+    /* TODO: check case where cap is "allow rw *". */
+
+    if (g.match.fs_name == ng.match.fs_name) {
+      if (g.match.path == ng.match.path and
```
@rishabh-d-dave I think these needs to be included during merge. Is that true?

rishabh-d-dave replied:
@vshankar PTAL - https://github.com/ceph/ceph/compare/330591f7ce19ab03b2e188d336a4e747d2d3cffd..a8104fdeb680be931a02493c3eb1d501000a87ba

vshankar commented:
@rishabh-d-dave So, the main pending item in this change is https://github.com/ceph/ceph/pull/41779#discussion_r1140948480

rishabh-d-dave commented:
Marking this PR DNM because it needs PR #50497 to be merged first.

vshankar commented:
> Marking this PR DNM because it needs PR #50497 to be merged first.

#50497 is good to merge. Pending item is: https://github.com/ceph/ceph/pull/41779#issuecomment-1480689281

vshankar commented:
@rishabh-d-dave Please fix the make check failure:

```
./tasks/cephfs/test_multifs_auth.py:7:1: F401 'tasks.cephfs.caps_helper.gen_mon_cap_str' imported but unused
./tasks/cephfs/test_multifs_auth.py:16:35: F821 undefined name 'CapsUtil'
flake8: exit 1 (4.06 seconds) /home/jenkins-build/build/workspace/ceph-pull-requests/qa> flake8 --select=F,E9 --exclude=venv,.tox pid=804034
flake8: FAIL ✖ in 16.48 seconds
```

rishabh-d-dave commented:
Rebased since PR #50497 has been merged.

rishabh-d-dave commented:
Reworded last commit message.

rishabh-d-dave commented:
Merged 2nd last and 3rd last commit into one since 2nd last was modifying code in tests according to PR #50497.

rishabh-d-dave commented on src/mds/MDSAuthCaps.cc:
```diff
@@ -373,6 +373,84 @@ bool MDSAuthCaps::parse(CephContext *c, std::string_view str, ostream *err)
+    if (g.match.fs_name == ng.match.fs_name) {
+      if (g.match.path == ng.match.path &&
+	  g.match.root_squash == ng.match.root_squash &&
+	  g.spec.get_caps() == ng.spec.get_caps()) {
```
Overloading operator `==` is not suitable idea here. I've talked with Greg about this. Marking this thread resolved.

rishabh-d-dave commented:
Major changes in last push -

* caught bug where client lost access to client path if `fs authorize` was run again with the same FS but a different path in arguments
* updated tests accordingly to this bug fix
* add new tests
* resolved all TODOs in tests
* refactored tests to reduce total LoC. Threw away around 100 LoC.
* rearranged all test classes and test methods in accordance with recent changes

rishabh-d-dave commented on src/mds/MDSAuthCaps.cc:
```diff
@@ -373,6 +373,80 @@ bool MDSAuthCaps::parse(CephContext *c, std::string_view str, ostream *err)
+
+  for (auto& g : grants) {
+    if (g.match.fs_name == ng.match.fs_name && g.match.path == ng.match.path) {
+      if (g.match.uid == ng.match.uid && g.match.gids == ng.match.gids &&
```
Next, tests caps for UID and GID myself, choose right behaviour of fs authorize in this case. Added tests and docs accordingly.

rishabh-d-dave replied:
Also added release note.

rishabh-d-dave commented:
Rebased to reolve conflicts.

rishabh-d-dave commented:
http://pulpito.front.sepia.ceph.com/rishabh-2023-04-11_05:58:23-fs-wip-rishabh-fs-auth-subcmd-2-distro-default-smithi/

vshankar commented:
@rishabh-d-dave I reviewing this today. Could you remove commits that have been carved out as a separate PR?

vshankar commented on src/mon/AuthMonitor.cc:
```diff
@@ -1795,6 +1795,47 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+
+    EntityAuth e_auth;
+    if (mon.key_server.get_auth(entity, e_auth)) {
+      for (auto it = newcaps.begin(); it != newcaps.end(); it += 2) {
```
This loop has been removed in commit 0e445f84616d812b6c2e3236d07c5b535d5cbc6b - why not just remove it in this commit?

rishabh-d-dave replied:
The loop wasn't added, it was just modified. 0e445f8 deals with a change different from c0d4aa8.

vshankar replied:
IMO, it the commits should be squashed. Can you explain why they are entirely different?

rishabh-d-dave replied:
Please take a look at the commit messages.

vshankar replied:
I already did, and it still looks pretty close enough to be treated as a single commit:

AuthMonitor: allow "fs authorize" to update caps
AuthMonitor: update caps when no caps are present in keyring

rishabh-d-dave replied:
> AuthMonitor: allow "fs authorize" to update caps

This commit adds the ability to add caps using `fs authorize` command

> AuthMonitor: update caps when no caps are present in keyring

And this commit changes the rule that empty keyring cannot be updated.

rishabh-d-dave replied:
@vshankar ping.

vshankar replied:
It still looks related, however, I'm not going to hold the progress for this change - so, I'm fine with having this as separate commits.

vshankar commented on src/mds/MDSAuthCaps.cc:
```diff
@@ -373,6 +373,89 @@ bool MDSAuthCaps::parse(CephContext *c, std::string_view str, ostream *err)
   }
 }
 
+bool MDSAuthCaps::merge(MDSAuthCaps newcap)
```
This method does not take into account uid/gid. I understand that you mentioned about the bug where uid/gid cannot be a part of the auth caps w/ fsname. So, will uid/gid be taken into account in that fix?

rishabh-d-dave replied:
`fs authorize` doesn't accept uid and gids as parameters. So there's no need for this method to consider them.

vshankar replied:
Sure. Please ensure its taken care in the PR where uid/gid can be part of the auth caps with file system name.

vshankar commented on qa/tasks/cephfs/test_admin.py:
```diff
@@ -1316,6 +1317,294 @@ def setup_test_env(self, fs_auth_caps):
+        that already had caps for first FS.
+        """
+        self.fs1 = self.fs
+        self.fs2 = self.mds_cluster.newfs('testcephfs2')
```
This test class has `MDSS_REQUIRED = 2` - were any warnings generated when a new file system was created (since there won't be any standby MDSs left)? If not, then is the warning in ignore-list set of warnings?

rishabh-d-dave replied:
Any warning generated in teuthology logs? No idea.

vshankar replied:
If the number of MDSs configured in the cluster is 2, then yes, a warning is generated. I think we run this test with >2 MDSs, so its fine, yes?

rishabh-d-dave replied:
> I think we run this test with >2 MDSs, so its fine, yes?

MDSS_REQUIRED is set to 2, so  this can't be true.

vshankar reviewed (commented):
Mostly looks fine. Nice work @rishabh-d-dave. Just a couple of points apart from comments:

- This change touches mon and osd cap - we should seek reviews from core team. Has it been reviewed by them? @rzarzynski
- Some of the commits are carved out as a separate PR. Please remove those commits on your next update @rishabh-d-dave

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1870,6 +1891,96 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+	return 2;
+      }
+
+      if (c_cap.merge(n_cap)) {
```
Done.

vshankar replied:
cant caps_update_reqd be a boolean?

rishabh-d-dave replied:
Didn't we change this to enum...

rishabh-d-dave replied:
I checked. We did.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1870,6 +1891,96 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+ * provided by the user at "fs authorize" command.
+ *
+ * Return values and their meaning
+ * 0 - caps update not required
```
Change has been made.

Copying for future reference - 

```

/* Generate the caps that should be present in the entity's auth keyring
 * after running the "fs authorize" command. This is done by merging the
 * caps already present in the client's auth keyring with the new caps
 * provided by the user at "fs authorize" command.
 *
 * Return values and their meaning
 * 0 - caps update not required
 * 1 - caps update is required
 * 2 - miscellanous error
 */
int AuthMonitor::_gen_wanted_caps(EntityAuth& e_auth,
```

vshankar commented:
@rishabh-d-dave commit 1559ef731579b7d8c3298f4c0a6f689aaa499570 is now another PR -  can you drop that from this change set?

rishabh-d-dave commented:
Return values of _gen_wanted_caps before we began using enum - https://github.com/ceph/ceph/pull/41779#discussion_r1172599742

rishabh-d-dave commented:
Rebased since a couple of related PRs (PR #50882 and PR #)51127) were merged recently

rishabh-d-dave commented:
jenkins test make check

vshankar commented:
> @vshankar
> 
> > @rishabh-d-dave commit [1559ef7](https://github.com/ceph/ceph/commit/1559ef731579b7d8c3298f4c0a6f689aaa499570) is now another PR - can you drop that from this change set?
> 
> Can we keep the commits that are copied to other PR? These commits make testing and debugging easy.

You can always create a branch with changes from multiple PRs for testing.

vshankar commented on src/mon/AuthMonitor.cc:
```diff
@@ -1795,6 +1795,47 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+      }
+
+      int rv = _gen_wanted_caps(e_auth, newcaps, &ss);
+      if (rv == 2) {
```
This should check the enumerated value.

vshankar replied:
By defining it in the header or even in the cc source in a scope that accessible to other class methods.

rishabh-d-dave replied:
You didn't mention this before. I was under impression that we want to use enum only inside _gen_wanted_caps().

vshankar replied:
Sorry, it was pretty obvious to me since if you define an enum for a function to return that enum, the callers should check against those enumerated value types.

rishabh-d-dave replied:
> I've made this change. PTAL.

 IMHO this getting much more complicated than using simple int. Especially, because this enum is not used beyond the very limited context of _gen_wanted_caps().

vshankar commented on src/mon/AuthMonitor.cc:
```diff
@@ -1870,6 +1891,80 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+  vector<string>& newcaps, ostream* out)
+{
+  enum caps_update_reqd_states {
+  NOT_REQD,	// caps need not be updated
```
How about:

```
CAP_UPDATE_NEEDED
CAP_UPDATE_SKIP
CAP_UPDATE_MISC_ERROR
```

rishabh-d-dave replied:
Better. But I am using `CAP_UPDATE_NOT_REQD` instead of `CAP_UPDATE_SKIP` since we are not skipping but it's  not needed.

vshankar commented on src/mon/AuthMonitor.cc:
```diff
@@ -1870,6 +1891,80 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+ * caps already present in the client's auth keyring with the new caps
+ * provided by the user at "fs authorize" command.
+ */
+int AuthMonitor::_gen_wanted_caps(EntityAuth& e_auth,
```
return should be enum type.

rishabh-d-dave replied:
Done. 

I'll say the same thing here too -- this much more complicated than simply using `int`; especially because this `enum` is not used beyond the very limited context of _gen_wanted_caps().

vshankar commented:
> > > @vshankar
> > > > @rishabh-d-dave commit [1559ef7](https://github.com/ceph/ceph/commit/1559ef731579b7d8c3298f4c0a6f689aaa499570) is now another PR - can you drop that from this change set?
> > > 
> > > 
> > > Can we keep the commits that are copied to other PR? These commits make testing and debugging easy.
> > 
> > 
> > You can always create a branch with changes from multiple PRs for testing.
> 
> For teuthology, yes. But testing locally (which is needed very often) becomes a mess with such an approach.

I'm worried that we might run into complications similar to https://github.com/ceph/ceph/pull/50882#discussion_r1169630886.

rishabh-d-dave commented:
@vshankar 
> > > > @vshankar
> > > > > @rishabh-d-dave commit [1559ef7](https://github.com/ceph/ceph/commit/1559ef731579b7d8c3298f4c0a6f689aaa499570) is now another PR - can you drop that from this change set?
> > > > 
> > > > 
> > > > Can we keep the commits that are copied to other PR? These commits make testing and debugging easy.
> > > 
> > > 
> > > You can always create a branch with changes from multiple PRs for testing.
> > 
> > 
> > For teuthology, yes. But testing locally (which is needed very often) becomes a mess with such an approach.
> 
> I'm worried that we might run into complications similar to [#50882 (comment)](https://github.com/ceph/ceph/pull/50882#discussion_r1169630886).

It won't be a problem because as those PRs are merged, these commit will disappear when I rebase this PR branch.

That was a different case -- testing branch had multiple PRs.

rishabh-d-dave commented on src/mds/MDSAuthCaps.h:
```diff
@@ -93,6 +93,17 @@ struct MDSCapSpec {
+    caps = _caps;
+  }
+
+  std::string str();
```
Done.

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
https://jenkins.ceph.com/job/ceph-pull-requests/114712/

rishabh-d-dave commented:
https://jenkins.ceph.com/job/ceph-pull-requests/114712/

rishabh-d-dave commented:
https://jenkins.ceph.com/job/ceph-pull-requests/114744/

rishabh-d-dave commented:
jenkins test make check

vshankar commented:
@rishabh-d-dave Please follow up with @rzarzynski for reviews. Once done, we can request runs from Yuri for all subsystems.

rishabh-d-dave commented:
Tests ran fine - https://pulpito.ceph.com/rishabh-2023-05-15_16:02:49-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

rishabh-d-dave commented:
jenkins test make check

vshankar commented:
> Tests ran fine - https://pulpito.ceph.com/rishabh-2023-05-15_16:02:49-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

@rishabh-d-dave I see you asked for re-review. I'll plan to do that today. If the MDS changes look fine then you can plan to run it through fs suite. However, we still need to wait for reviews from @rzarzynski.

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
> > Tests ran fine - https://pulpito.ceph.com/rishabh-2023-05-15_16:02:49-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/
> 
> @rishabh-d-dave I see you asked for re-review. I'll plan to do that today. If the MDS changes look fine then you can plan to run it through fs suite. However, we still need to wait for reviews from @rzarzynski.

ping

vshankar commented:
> > > Tests ran fine - https://pulpito.ceph.com/rishabh-2023-05-15_16:02:49-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/
> > 
> > 
> > @rishabh-d-dave I see you asked for re-review. I'll plan to do that today. If the MDS changes look fine then you can plan to run it through fs suite. However, we still need to wait for reviews from @rzarzynski.
> 
> ping

I have no review comments atm. This required review from core team which I guess will be prioritised soon.

rzarzynski commented:
Review-in-progress.

rzarzynski commented on src/mds/MDSAuthCaps.cc:
```diff
@@ -373,6 +373,89 @@ bool MDSAuthCaps::parse(CephContext *c, std::string_view str, ostream *err)
 
+bool MDSAuthCaps::merge(MDSAuthCaps newcap)
+{
+  assert(newcap.grants.size() == 1);
```
I don't know the `mds` code base and its conventions. In other parts typically `ceph_assert()` is used. Is `assert()` intentional here?

rishabh-d-dave replied:
I talked with Patrick about. It appears that using `ceph_assert()` is better.

rzarzynski commented on src/mds/MDSAuthCaps.h:
```diff
@@ -93,6 +93,17 @@ struct MDSCapSpec {
+    return caps;
+  }
+
+  void set_caps(unsigned _caps) {
```
UUIC, `unsigned` alone is discouraged in new code but `MDSCapSpec::caps` is already typed that way.

rishabh-d-dave replied:
Done.

rzarzynski reviewed (commented):
Part 1/2 -- initial commits. At the moment just nits. Going over `AuthMonitor`.

rzarzynski reviewed (approved):
Part 2/2. LGTM with some tiny nits / questions.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1921,6 +1899,39 @@ void AuthMonitor::_print_key(EntityName& entity, EntityAuth& eauth,
   _print_auth(entity, eauth, rdata, fmtr, true, wanted_caps);
 }
 
+int AuthMonitor::_update_caps(EntityName& entity, vector<string>& caps_vec,
```
Done/

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1843,33 +1843,11 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
-    if (!valid_caps(caps_vec, &ss)) {
-      err = -EINVAL;
+    err = _update_caps(entity, caps_vec, op);
+    if (err == 0)
```
Added.

rishabh-d-dave commented on src/mds/MDSAuthCaps.cc:
```diff
@@ -373,6 +373,89 @@ bool MDSAuthCaps::parse(CephContext *c, std::string_view str, ostream *err)
+{
+  string str = "";
+
+  if (!fs_name.empty())   { str += " fsname=" + fs_name; }
```
> I'm assuming it's fine to have the string always beginning with a space

It is because `fsname=<fs-name>` always appears later in MDS caps.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1870,6 +1891,91 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+    string c_cap_str; // current cap held by entity's auth keyring
+    auto iter = e_auth.caps[cap_entity].cbegin();
+    decode(c_cap_str, iter);
+    if (c_cap_str == n_cap_str) {
```
Agreed. Switched to `new_` and `cur_` instead.

rishabh-d-dave commented:
.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1808,26 +1849,6 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
-    }
-
     KeyServerData::Incremental auth_inc;
     auth_inc.op = KeyServerData::AUTH_INC_ADD;
```
In this case we are creating an new entity, not updating its caps. But most LOC for both are same for both so I made some modifications and got this done.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1870,6 +1891,91 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+    return CAPS_UPDATE_REQD;
+  }
+
+  for (auto it = newcaps.begin(); it != newcaps.end(); it += 2) {
```
This change shouldn't be necessary once I replace this vector by map and I iterate over like this: `for (auto it : map)`.

Here's your comment where you requested me to replace the vector with map - https://github.com/ceph/ceph/pull/41779#discussion_r1205591883

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1870,6 +1891,91 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+ * provided by the user at "fs authorize" command.
+ */
+AuthMonitor::caps_update AuthMonitor::_gen_wanted_caps(EntityAuth& e_auth,
+  vector<string>& newcaps, ostream* out)
```
Done.

rishabh-d-dave commented:
.

rishabh-d-dave commented:
.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1795,6 +1795,47 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+    if (mon.key_server.get_auth(entity, e_auth)) {
+      for (auto it = newcaps.begin(); it != newcaps.end(); it += 2) {
+	string& cap_entity = *it;
+        if (entity_auth.caps.count(cap_entity) == 0) {
```
As discussed on chat, this message is deleted in following commits.

rishabh-d-dave replied:
Marking this conversation resolved.

rzarzynski commented on src/mon/AuthMonitor.cc:
```diff
@@ -1941,6 +1898,69 @@ void AuthMonitor::_encode_key(const EntityName& entity,
+
+int AuthMonitor::_create_entity(const EntityName& entity,
+  const vector<string>& caps_vec, MonOpRequestRef op, stringstream& ds,
+  bufferlist* rdata, Formatter* fmtr) {
```
nit: let's stick to the local, already established convention for the opening brace (`{`).

rishabh-d-dave replied:
Sure.

I found that this local convention, of placing opening brace on a newline, brings more space and therefore makes code more readable, especially when a function has many parameters.

rishabh-d-dave replied:
Done.

rishabh-d-dave replied:
Done.

rzarzynski commented on src/mon/AuthMonitor.cc:
```diff
@@ -1801,19 +1801,6 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
-
     EntityAuth entity_auth;
     if (mon.key_server.get_auth(entity, entity_auth)) {
       for (const auto &sys_cap : wanted_caps) {
```
Which `wanted_caps` do we use here after removing the `wanted_caps` definition several lines above?

rzarzynski commented on src/mon/AuthMonitor.cc:
```diff
@@ -1801,24 +1801,44 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+
     EntityAuth entity_auth;
     if (mon.key_server.get_auth(entity, entity_auth)) {
-      for (const auto &sys_cap : wanted_caps) {
```
OK, this is the response for the question on `wanted_caps`.

rzarzynski reviewed (approved):
LGTM with some minor questions / nits.

rishabh-d-dave commented:
Tests ran fine - https://pulpito.ceph.com/rishabh-2023-06-09_16:37:36-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1933,6 +1899,48 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
   return false;
 }
 
+void AuthMonitor::_encode_keyring(KeyRing& kr, const EntityName& entity,
```
What confusion? I find this code pretty simple. Perhaps I am overlooking something.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1390,7 +1381,7 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
   auto m = op->get_req<MMonCommand>();
   stringstream ss, ds;
-  bufferlist rdata;
+  bufferlist rdata; // holds data that'll be printed on stdout
```
Done.

rishabh-d-dave commented:
Tests ran fine this time as well - https://pulpito.ceph.com/rishabh-2023-06-10_12:13:14-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

batrick commented on doc/cephfs/client-auth.rst:
```diff
@@ -259,3 +259,118 @@ Following is an example of enabling root_squash in a filesystem except within
+    $ ceph auth get client.x
+    [client.x]
+            key = AQAOtSVk9WWtIhAAJ3gSpsjwfIQ0gQ6vfSx/0w==
+            caps mds = "allow rw fsname=a, allow rw fsname=a path=some/dir"
```
This second set of caps is redundant. It'd be better if:

```suggestion
            caps mds = "allow r fsname=a, allow rw fsname=a path=some/dir"
```

rishabh-d-dave replied:
Fixed.

batrick commented on doc/cephfs/client-auth.rst:
```diff
@@ -259,3 +259,118 @@ Following is an example of enabling root_squash in a filesystem except within
+Running second command can be interpreted as changing ``/dir1`` to ``/dir2``
+in current cap or can also be interpreted as authorizing the client with a
+new cap for path ``/dir2``. As seen in previous sections, second
+interpretation is chosen and thefore it's impossible to update a part of
```
```suggestion
interpretation is chosen and therefore it's impossible to update a part of
```

rishabh-d-dave replied:
Fixed.

batrick commented on doc/cephfs/client-auth.rst:
```diff
@@ -259,3 +259,118 @@ Following is an example of enabling root_squash in a filesystem except within
+            key = AQBBKjBkIFhBDBAA6q5PmDDWaZtYjd+jafeVUQ==
+            caps mds = "allow r fsname=a"
+            caps mon = "allow r fsname=a"
+            caps osd = "allow r tag cephfs data=a"
```
These examples would be better if they were not interdependent. Please make each example standalone.

rishabh-d-dave replied:
IIUC you want me to show the client creation in example along with client update. I avoided that because each example size would become double.

batrick replied:
I think that's fine. You don't need to show the client cap when first settign it up. IT would be something like:

```
$ ceph fs authorize a client.x / rw
$ ceph fs authorize a client.x / r
$ ceph auth get client.x
...
```

You can omit output if it's not helpful for understanding.

rishabh-d-dave replied:
IMO, a user needs to see the output because only the output reflects impact of running `fs authorize` command in succession.

batrick commented on doc/cephfs/client-auth.rst:
```diff
@@ -259,3 +259,118 @@ Following is an example of enabling root_squash in a filesystem except within
+running ``fs authorize`` again for FS name ``a`` but path ``dir2``, instead
+of modifying the caps client already holds, a new cap for ``dir2`` will be
+granted::
+
```
Perhaps there should be some `fs deauthorize`? Let's make a ticket for further enhancement?

batrick replied:
A lot of that ambiguity also exists for `fs authorize`. I think you should defer that to a future enhancement because this PR has gone on long enough.

rishabh-d-dave replied:
Yes, but should I defer it to near future, start working on it parallely and create a new PR for it (I originally had this in my mind) or postpone indefinitely? From my conversation with Greg, my impression is that I should postpone it indefinitely, at least for now.

batrick replied:
Suggest discussing with Venky/Greg.

rishabh-d-dave replied:
Thanks for the replies (here as well as on chat).

rishabh-d-dave commented:
Rebased the PR, resolved the merge conflicts, and made following changes to AuthMonitor -

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1933,6 +1843,196 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
   return false;
 }
 
+template<typename CAP_ENTITY_CLASS>
+AuthMonitor::caps_update AuthMonitor::_merge_caps(const string& cap_entity,
+  const string& new_cap_str, const string& cur_cap_str,
+  map<string, string>& newcaps, ostream& out)
+{
+  CAP_ENTITY_CLASS cur_cap, new_cap;
+
+  if (not cur_cap.parse(cur_cap_str, &out)) {
+    out << "error parsing " << cap_entity << "caps client already holds";
+    return CAPS_PARSING_ERR;
+  }
+  if (not new_cap.parse(new_cap_str, &out)) {
+    out << "error parsing new " << cap_entity << "caps";
+    return CAPS_PARSING_ERR;
+  }
+
+  if (cur_cap.merge(new_cap)) {
+    newcaps[cap_entity] = cur_cap.to_string();
+    return CAPS_UPDATE_REQD;
+  } else {
+    newcaps[cap_entity] = cur_cap_str;
+    return CAPS_UPDATE_NOT_REQD;
+  }
+}
```
.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1933,6 +1843,196 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+      continue;
+    }
+
+    if (cap_entity == "mon") {
+      is_caps_update_reqd = _merge_caps<MonCap>(cap_entity, new_cap_str,
+	cur_cap_str, newcaps, out);
+    } else if (cap_entity == "osd") {
+      is_caps_update_reqd = _merge_caps<OSDCap>(cap_entity, new_cap_str,
+	cur_cap_str, newcaps, out);
+    } else if (cap_entity == "mds") {
+      is_caps_update_reqd = _merge_caps<MDSAuthCaps>(cap_entity, new_cap_str,
+	cur_cap_str, newcaps, out);
```
.

rishabh-d-dave commented:
.

rishabh-d-dave commented:
Rebased and resolved conflicrs since https://github.com/ceph/ceph/pull/52008 was merged.

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
Tests ran fine -

rishabh-d-dave commented:
https://jenkins.ceph.com/job/ceph-api/57479/

rzarzynski commented on src/mon/MonCap.cc:
```diff
@@ -690,3 +690,62 @@ bool MonCap::parse(const string& str, ostream *err)
+
+  for (size_t i = 0; i < grants.size(); ++i) {
+    str += grants[i].to_string();
+    if (i < grants.size () - 1) {
```
OK, for empty `grants` there would be no iteration.

rzarzynski reviewed (approved):
The `AuthMonitor`-related changes LGTM with some nits.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1842,6 +1861,76 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+    return CAPS_UPDATE_REQD;
+  }
+
+  for (auto it : newcaps) {
```
Yeah, that'll be much better. Done.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1789,9 +1789,11 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
-      for (const auto &sys_cap : encoded_caps) {
-	if (entity_auth.caps.count(sys_cap.first) == 0 ||
-	    !entity_auth.caps[sys_cap.first].contents_equal(sys_cap.second)) {
+      for (const auto& it : encoded_caps) {
```
Much better. Done.

rishabh-d-dave commented on src/mon/AuthMonitor.cc:
```diff
@@ -1802,9 +1804,26 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
+	ss << "no update for caps of " << entity;
+	err = 0;
+	goto done;
+      }
```
That'll be safer.  Done.

rishabh-d-dave commented on src/mds/MDSAuthCaps.cc:
```diff
@@ -371,6 +371,89 @@ bool MDSAuthCaps::parse(std::string_view str, ostream *err)
+    str += " gids=";
+    for (size_t i = 0; i < gids.size(); ++i) {
+      str += std::to_string(gids[i]);
+      if (i < gids.size() - 1) {
```
No need for parentheses on which line?

rzarzynski replied:
```
      if (i < gids.size() - 1) {
```

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
https://jenkins.ceph.com/job/ceph-api/57965/

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
https://jenkins.ceph.com/job/ceph-api/57965/

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
Tests ran fine - https://pulpito.ceph.com/rishabh-2023-07-12_06:51:12-fs-wip-rishabh-fs-auth-subcmd-distro-default-smithi/

(The one failure on the above run occurs after the testing was complete)

rzarzynski commented:
I think with an approval from CephFS we're ready for `needs-qa`.

vshankar reviewed (commented):
LGTM

rishabh-d-dave reviewed (commented):
Tests ran successfully - https://tracker.ceph.com/projects/cephfs/wiki/Main#24-Jul-2024

rishabh-d-dave commented:
jenkins test docs

rishabh-d-dave commented:
The patch for PendingReleaseNote needs an update. I'll make it and then merge the PR.

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
jenkins test docs

rishabh-d-dave commented:
Just for record: bugs introduced into this PR and on main branch were was introduced on PR #50569.

ljflores commented:
@rishabh-d-dave do you want this in one of Yuri's batches? I see you added your own testing label, so wanted to confirm.

rishabh-d-dave commented:
> @rishabh-d-dave do you want this in one of Yuri's batches? I see you added your own testing label, so wanted to confirm.

Do you mean put it through RADOS integration tests?

rishabh-d-dave commented:
@ljflores 
> @rishabh-d-dave do you want this in one of Yuri's batches? I see you added your own testing label, so wanted to confirm.

Do you mean testing this PR with RADOS integration tests?

ljflores commented:
@rishabh-d-dave yes, do you need us to put it through a rados batch? Either way, it should go through the rados suite since it has a "core" label. It's just a question whether you are already doing it, or if you want us to do it through Yuri's pipeline.

ljflores commented:
jenkins retest this please

ljflores commented:
@rishabh-d-dave I went ahead and put it in one of Yuri's batches. Feel free to continue with your own testing though if preferred.

rishabh-d-dave commented:
@ljflores 
> @rishabh-d-dave I went ahead and put it in one of Yuri's batches. Feel free to continue with your own testing though if preferred.

Thanks. Sorry for no reply from my end. I was super-busy last week due to some non-office work. I think this PR had gone through RADOS testing before, but I am not fully sure. Please let me know when the run is complete and results are out.

ljflores commented:
Hi @rishabh-d-dave, tests have completed, so a review will be ready soon.

ljflores commented:
jenkins retest this please

ljflores commented:
@rishabh-d-dave you can monitor progress here: https://trello.com/c/Fllj7bVM/1833-wip-yuri8-testing-2023-08-28-1340

rishabh-d-dave commented:
jenkins test docs

rishabh-d-dave commented:
This failure message looks legit. Once this review of integration test run is posted, I'll fix it.
```
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/cephfs/client-auth.rst:286: WARNING: Literal block expected; none found.
```

rishabh-d-dave reviewed (commented):
CephFS integration tests ran fine - https://tracker.ceph.com/projects/cephfs/wiki/Main#6-Sep-2023.

ljflores commented:
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcFllj7bVM1833-wip-yuri8-testing-2023-08-28-1340

ljflores commented:
Feel free to merge when the docs issue is fixed.

rishabh-d-dave commented:
Fixed a minor doc issue.

Waiting for CI to be green, after this PR can be merged since both CephFS and RADOS tests ran successfully.

rishabh-d-dave commented:
Minor doc update.

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
jenkins test api
